packages/multi-test/src/contracts.rs

```diff
@@ -0,0 +1,267 @@
+type ReplyClosure<C, E> = Box<dyn Fn(DepsMut, Env, Reply) -> Result<Response<C>, E>>;
+type QueryClosure<T, E> = Box<dyn Fn(Deps, Env, T) -> Result<Binary, E>>;
+
+/// Wraps the exported functions from a contract and provides the normalized format
+/// Place T4 and E4 at the end, as we just want default placeholders for most contracts that don't have sudo
+pub struct ContractWrapper<T1, T2, T3, E1, E2, E3, C = Empty, T4 = Empty, E4 = String, E5 = String>
+where
+    T1: DeserializeOwned,
+    T2: DeserializeOwned,
+    T3: DeserializeOwned,
+    T4: DeserializeOwned,
+    E1: ToString,
+    E2: ToString,
+    E3: ToString,
+    E4: ToString,
+    E5: ToString,
+    C: Clone + fmt::Debug + PartialEq + JsonSchema,
+{
+    execute_fn: ContractClosure<T1, C, E1>,
+    instantiate_fn: ContractClosure<T2, C, E2>,
+    query_fn: QueryClosure<T3, E3>,
+    sudo_fn: Option<SudoClosure<T4, C, E4>>,
+    reply_fn: Option<ReplyClosure<C, E5>>,
+}
+
+impl<T1, T2, T3, E1, E2, E3, C> ContractWrapper<T1, T2, T3, E1, E2, E3, C>
+where
+    T1: DeserializeOwned + 'static,
+    T2: DeserializeOwned + 'static,
+    T3: DeserializeOwned + 'static,
+    E1: ToString + 'static,
+    E2: ToString + 'static,
+    E3: ToString + 'static,
+    C: Clone + fmt::Debug + PartialEq + JsonSchema + 'static,
+{
+    pub fn new(
+        execute_fn: ContractFn<T1, C, E1>,
+        instantiate_fn: ContractFn<T2, C, E2>,
+        query_fn: QueryFn<T3, E3>,
+    ) -> Self {
+        ContractWrapper {
+            execute_fn: Box::new(execute_fn),
+            instantiate_fn: Box::new(instantiate_fn),
+            query_fn: Box::new(query_fn),
+            sudo_fn: None,
+            reply_fn: None,
+        }
+    }
+
+    /// this will take a contract that returns Response<Empty> and will "upgrade" it
+    /// to Response<C> if needed to be compatible with a chain-specific extension
+    pub fn new_with_empty(
+        execute_fn: ContractFn<T1, Empty, E1>,
+        instantiate_fn: ContractFn<T2, Empty, E2>,
+        query_fn: QueryFn<T3, E3>,
+    ) -> Self {
+        ContractWrapper {
+            execute_fn: customize_fn(execute_fn),
+            instantiate_fn: customize_fn(instantiate_fn),
+            query_fn: Box::new(query_fn),
+            sudo_fn: None,
+            reply_fn: None,
+        }
+    }
+}
+
+impl<T1, T2, T3, E1, E2, E3, C, T4, E4, E5> ContractWrapper<T1, T2, T3, E1, E2, E3, C, T4, E4, E5>
+where
+    T1: DeserializeOwned + 'static,
+    T2: DeserializeOwned + 'static,
+    T3: DeserializeOwned + 'static,
+    T4: DeserializeOwned + 'static,
+    E1: ToString + 'static,
+    E2: ToString + 'static,
+    E3: ToString + 'static,
+    E4: ToString + 'static,
+    E5: ToString,
+    C: Clone + fmt::Debug + PartialEq + JsonSchema + 'static,
+{
+    pub fn with_sudo<T4A, E4A>(
+        self,
+        sudo_fn: SudoFn<T4A, C, E4A>,
+    ) -> ContractWrapper<T1, T2, T3, E1, E2, E3, C, T4A, E4A, E5>
```

Comment:
I guess there's no way around this, because of the defaults. Perhaps without specifying the defaults? They are used to specify a unset `Option` in the end.

Reply:
Way around what?

Changing T4, E4 -> T4A, E4A?

I was quite happy with this discovery, as all attempts to use `&mut self` and change the struct failed (as mis-matched types), but the builder allows us to customise the types as we wish.